bpo-42197: Don't create `f_locals` dictionary unless we actually need it. (#32055)
python/cpython: Misc/NEWS.d/next/Core and Builtins/2022-03-22-15-12-28.bpo-42197.SwrrFO.rst

:c:func:`PyFrame_FastToLocalsWithError` and :c:func:`PyFrame_LocalsToFast` are no longer

> called during profiling nor tracing. C code can access the ``f_locals`` attribute of :c:type:`PyFrameObject` by calling :c:func:`PyFrame_GetLocals`.

Reviewer: Please copy this paragraph to What's New In Python 3.11 > C API > Porting to Python 3.11:
https://docs.python.org/dev/whatsnew/3.11.html#id2

By the way, in 2013, it was decided to **not** document these functions: https://bugs.python.org/issue19431

Maybe we should revisit this decision? What do you think? I can document them in the new frame.rst doc if you want.

Author: There should be no need to document this is the porting to 3.11 section. It should "just work".
The important change C code must make is to call `PyFrame_GetLocals()` instead of access the `f_locals` field directly. That is already mentioned here and in the whatsnew file.

Reviewer: Developers of profilers and debuggers (implemented in C) will have to dig into Changelog or even the Python bug tracker to understand what's going on. Since it's a behavior change, it should be documented.

I agree that profilers and debuggers implemented in pure Python will not be affected. And people not using profiling or debugging will not be affected :-)

Author: OK, I've added a section to the whatsnew

> ``PyFrame_FastToLocalsWithError`` and ``PyFrame_LocalsToFast`` are no longer
> called during profile nor tracing. C code that accesses the ``f_locals`` attribute
>  directly must do so by calling ``PyFrame_GetLocals``.

Reviewer 2: ```suggestion
:c:func:`PyFrame_FastToLocalsWithError` and :c:func:`PyFrame_LocalsToFast` are no longer
called during profiling nor tracing. C code can access the ``f_locals`` attribute of :c:type:`PyFrameObject` by calling :c:func:`PyFrame_GetLocals`.
```

``PyFrame_FastToLocalsWithError`` and ``PyFrame_LocalsToFast`` are no longer
called during profile nor tracing. C code that accesses the ``f_locals`` attribute
 directly (which is not recommended) must now call
``PyFrame_FastToLocalsWithError`` to update locals and ``PyFrame_LocalsToFast``

> after ``f_locals`` are updated.

Reviewer: For this change, please an entry to What's New in Python 3.11 > C API > Porting to Python 3.11:
https://docs.python.org/dev/whatsnew/3.11.html#id2
